### PR TITLE
Use aligned block sync in block reduction runtime kernel

### PR DESCRIPTION
**Pull request by naoyam**
This PR changes how block sync is done in some of the runtime functions. Presently, we always use `__barrier_sync(0)` as runtime functions may be called with divergent threads, which isn't optimal if threads are aligned. This PR extends what we already do for the `BlockSync` IR node and uses `__syncthreads()` when possible.

The primary change is:

```
template <bool aligned = false>
__forceinline__ __device__ void sync() {
  if constexpr (aligned) {
    __syncthreads();
  } else {
    __barrier_sync(0);
  }
}
```

And for now I only extended `blockReduce` to use the `aligned` template parameter. 

Another major change is in codegen.cpp. I added `std::vector<bool> aligned_scope_exprs_` to represent if a scope is aligned or not. I thought maybe we might want to add a bool property to `Expr` to indicate if it's aligned or not, but it depends on the surrounding scopes, so adding that property to `Expr` doesn't seem like a good idea.

As part of extending `CudaKernelGenerator`, I also refactored it to use `kir::IrVisitor`. 

This PR only changes `blockReduce`. I'll have follow-up PRs to address other runtime routines.

In our current benchmarks, there are about 300 cases of `blockReduce`, of which about 200 are aligned. I'll get performance results as well.

Looks like there's no noticeable perf difference. The below graph compares the bandwidths of all the benchmarks that now use the aligned sync vs. the same benchmarks with the default barrier sync. There are 132 benchmarks that contain aligned blockReduce.

image

### CODE REVIEW

**Comment by naoyam** [comment]
!build

File: runtime/block_reduction.cu [hunk 1]
```diff
@@ -15,46 +15,38 @@
...
-//
 template <
     bool X_REDUCE,
     bool Y_REDUCE,
     bool Z_REDUCE,
+    bool aligned,
     typename T,
-    typename Func,
-    typename _dim3,
-    typename _dim3_2>
```
Comment (naoyam) [inline runtime/block_reduction.cu:30]: Not necessary anymore.

File: runtime/block_reduction.cu [hunk 2]
```diff
@@ -15,46 +15,38 @@
 //  blockReduceSum<X_THREADS, Y_THREADS, Z_THREADS>
 //    (output[output_index], inputs[input_index],
 //      [] __device__ (T& a, const T b) { a += b; });
-//
```
Comment (naoyam) [inline runtime/block_reduction.cu:18]: In addition to adding `aligned` template parameter, I also removed code that was only necessary for ROCM compatibility.

File: csrc/device_lower/lower2device.cpp [hunk 3]
```diff
@@ -160,7 +160,6 @@ class ConvertAlignedBlockSync : kir::IrVisitor {
                 RegisterType::GeneralPurpose) {
           return;
         }
-        return;
```
Comment (naoyam) [inline csrc/device_lower/lower2device.cpp:163]: @zasdfgbnm I believe this should not be here, right?

Comment (naoyam) [reply csrc/device_lower/lower2device.cpp:163]: I'll remove this function as I added a copy in ir_utils. Just want to make sure this `return` is not necessary.

Comment (zasdfgbnm) [reply csrc/device_lower/lower2device.cpp:163]: I think your understanding is correct. This fix do change behavior, and fortunately we don't see perf regression on it :)

Comment (naoyam) [reply csrc/device_lower/lower2device.cpp:163]: Thanks for confirmation. I'll do a quick cleanup

File: runtime/block_reduction.cu (hunk 1 above)
Comment (zasdfgbnm) [reply runtime/block_reduction.cu:30]: Was this only for ROCm? This looks annoying and make the code unreadable. It's good that we can remove it now.

File: csrc/device_lower/lower2device.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply csrc/device_lower/lower2device.cpp:163]: Are you planning to remove this in this PR, or do it in a separate PR?

Comment (naoyam) [reply csrc/device_lower/lower2device.cpp:163]: Done. Running the tests locally just in case

File: csrc/codegen.cpp [hunk 4]
```diff
@@ -1098,30 +1128,31 @@ class CudaKernelGenerator : private OptOutConstDispatch {
...
-             << (tidy ? "true" : "false") << ", " << (tidz ? "true" : "false")
-             << ">(\n";
-    indent() << kTab << gen(output) << ",\n";
-    indent() << kTab << gen(input) << ",\n";
-    indent() << kTab << genReductionOp(reduction_op_type, output->dtype())
-             << ",\n";
-    indent() << kTab << "threadIdx,\n";
-    indent() << kTab << "blockDim,\n";
-    indent() << kTab << "static_cast<" << data_type << "*>(shared_mem),\n";
+    ArgumentBuilder template_args;
```
Comment (zasdfgbnm) [inline csrc/codegen.cpp:1131]: nice cleanup!

File: csrc/codegen.cpp [hunk 5]
```diff
@@ -2680,7 +2709,7 @@ class CudaKernelGenerator : private OptOutConstDispatch {
     }
   }
 
-  void handle(const kir::BlockSync* sync) final {
+  void handle(kir::BlockSync* sync) final {
     // Use a custom synchronization method if enabled
     if (std::getenv("PYTORCH_NVFUSER_USE_BLOCK_SYNC_ATOMIC")) {
       indent() << "block_sync::sync();\n";
```
Comment (zasdfgbnm) [inline csrc/codegen.cpp:2715]: Does it make sense to do a further cleanup of `BlockSync` to remove `isAligned` from it, and remove `ConvertAlignedBlockSync` completely, and here just generate `f'block_sync::sync<{isAligned()}>()'`?

File: csrc/codegen.cpp [hunk 6]
```diff
@@ -2798,6 +2827,8 @@ class CudaKernelGenerator : private OptOutConstDispatch {
   std::deque<const kir::ForLoop*> grouped_loops_;
   //! Used to replace symbolic indices with concrete values
   std::unordered_map<const Int*, int64_t> index_replacement_map_;
+  //! Keep track of thread alignment property
+  std::deque<bool> aligned_scope_exprs_;
```
Comment (zasdfgbnm) [inline csrc/codegen.cpp:2831]: Why this is not a `vector`?

Comment (naoyam) [reply csrc/codegen.cpp:2831]: Just to do pop and push

File: runtime/block_reduction.cu (hunk 1 above)
Comment (naoyam) [reply runtime/block_reduction.cu:30]: Yes, it was for the compatibility

File: csrc/codegen.cpp (hunk 6 above)
Comment (zasdfgbnm) [reply csrc/codegen.cpp:2831]: I think you only used `push_back` and `pop_back`? For this case, `vector` is better?

File: csrc/codegen.cpp (hunk 5 above)
Comment (zasdfgbnm) [reply csrc/codegen.cpp:2715]: I think so, the only reason why we do the conversion was to get better matmul perf, I think this goal can still be achieved with the suggest change.

File: csrc/codegen.cpp (hunk 6 above)
Comment (naoyam) [reply csrc/codegen.cpp:2840]: Isn't it better to use a linked list-like structure?

File: csrc/codegen.cpp [hunk 7]
```diff
@@ -120,7 +120,7 @@ std::string genCall(
   return ss.str();
 }
 
-class CudaKernelGenerator : private OptOutConstDispatch {
+class CudaKernelGenerator : private kir::IrVisitor {
```
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/codegen.cpp]: Changed. Not quite ideal as, for example, `IfThenElse::thenBody().exprs()` returns a vector of `Expr*` not `const Expr*`. We could add something like `IfThenElse::thenBody().constExprs()`, but this time I just added `handle(std::vector<Expr*>&)`

File: csrc/codegen.cpp (hunk 5 above)
Comment (naoyam) [reply csrc/codegen.cpp:2724]: Done

**Comment by naoyam** [comment]
!build

File: csrc/codegen.cpp (hunk 6 above)
Comment (zasdfgbnm) [reply csrc/codegen.cpp:2840]: Really? I think for vector, in most cases, a `push_back` is just a `storage[size++] = item`, and for linked-list like data structure, you will need to create a node and manipulate two pointers? I find this https://baptiste-wicht.com/posts/2012/12/cpp-benchmark-vector-list-deque.html

Comment (naoyam) [reply csrc/codegen.cpp:2840]: Thanks. That makes sense. I don't expect there's much performance difference as it shouldn't be a bottleneck, but I'll change it to use vector.

File: csrc/codegen.cpp [hunk 8]
```diff
@@ -1038,8 +1077,8 @@ class CudaKernelGenerator : private OptOutConstDispatch {
   }
 
   void genSerialReduction(
-      const kir::TensorIndex* output,
```
Comment (zasdfgbnm) [inline csrc/codegen.cpp:1041]: Should the `const` in this function be changed back? Similar for other functions.

**Review by zasdfgbnm — APPROVED** [review]
LGTM, the only final thing is https://github.com/NVIDIA/Fuser/pull/351#discussion_r1196955163

File: csrc/codegen.cpp (hunk 8 above)
Comment (naoyam) [reply csrc/codegen.cpp:1041]: Thanks. I thought I added back.

Comment (naoyam) [reply csrc/codegen.cpp]: Done

**Pull request merged**